File: Dynamic Programming/minimizing-coins/iterative_dp.cpp
#include <bits/stdc++.h>
using namespace std;

const long long inf = 1e10;

int main() {
	long long n, target;
	cin >> n >> target;
	long long coin;
	vector<long long> coins;
	for(int i = 0; i < n; i++) {
		cin >> coin;
		coins.push_back(coin);
	}
	vector<vector<long long >> dp(2, vector<long long>(target+1, inf));
	dp[0][0] = 0;
	int idx;
	for(long long i = 1; i < n+1; i++) {
		idx = i & 1;
		for(long long j = 0; j < target+1; j++) {
			long long pick = j - coins[i-1] >= 0 ? 1 + dp[idx][j - coins[i-1]] : inf;
			long long not_pick = dp[1 - idx][j];
			dp[idx][j] = min(pick, not_pick);
		}
	}
	long long ans = dp[idx][target];
	if(ans == inf) ans = -1;
	cout << ans << "\n";
} 

File: Dynamic Programming/minimizing-coins/iterative_dp.py
import resource, sys
resource.setrlimit(resource.RLIMIT_STACK, (2**29,-1))

inf = 1e10
n, target = map(int, input().split())
coins = list(map(int, input().split()))
dp = [[0] * (target+1) for _ in range(n+1)]

for i in range(n+1):
	for j in range(target+1):
		if i == 0 and j > 0:
			dp[i][j] = inf
			continue
		pick = 1 + dp[i][j - coins[i-1]] if j - coins[i-1] >= 0 else inf
		not_pick = dp[i-1][j]
		dp[i][j] = min(pick, not_pick)

result = dp[n][target]
print(result if result != inf else -1)